File: backend-cs/Services/FanTestService.cs
using DriveChill.Hardware;
using DriveChill.Models;

namespace DriveChill.Services;

/// <summary>
/// Manages fan benchmark tests — one test per fan, concurrent across different fans.
///
/// Each test sweeps speed from 0% → 100% in configurable steps, waits for RPM to
/// settle at each step, then records the result. Progress is streamed to WebSocket
/// clients via FanTestProgress objects in WsMessage.fan_test.
/// </summary>
public sealed class FanTestService
{
    private readonly FanService           _fans;
    private readonly SensorService        _sensors;
    private readonly IHardwareBackend     _hw;
    private readonly ILogger<FanTestService> _log;

    private readonly Dictionary<string, TestRun> _runs = new();
    private readonly object _lock = new();

    public FanTestService(FanService fans, SensorService sensors,
        IHardwareBackend hw, ILogger<FanTestService> log)
    {
        _fans    = fans;
        _sensors = sensors;
        _hw      = hw;
        _log     = log;
    }

    // -----------------------------------------------------------------------
    // Public API
    // -----------------------------------------------------------------------

    /// <summary>
    /// Starts a benchmark sweep for <paramref name="fanId"/>.
    /// Returns false (with a reason in <paramref name="error"/>) if the fan is not
    /// found or a test is already running for that fan.
    /// </summary>
    public bool TryStart(string fanId, FanTestOptions options, out string error)
    {
        if (!_hw.GetFanIds().Contains(fanId))
        {
            error = $"Fan '{fanId}' not found";
            return false;
        }

        lock (_lock)
        {
            if (_runs.TryGetValue(fanId, out var existing) && existing.Result.Status == "running")
            {
                error = $"A test is already running for fan '{fanId}'";
                return false;
            }

            // Capture current fan mode so we can restore it after the test
            var snapshot       = _sensors.Latest;
            var fanStatuses    = _fans.GetAll(snapshot);
            var currentStatus  = fanStatuses.FirstOrDefault(f => f.FanId == fanId);
            var previousMode   = currentStatus?.Mode ?? "auto";
            var previousCurve  = _fans.GetCurves().FirstOrDefault(c => c.FanId == fanId);

            var result = new FanTestResult
            {
                FanId   = fanId,
                Options = options,
            };

            var run = new TestRun(result);
            _fans.LockForTest(fanId);
            run.Task = Task.Run(() =>
                RunSweepAsync(fanId, run, previousMode, previousCurve, run.Cts.Token));

            _runs[fanId] = run;
        }

        error = "";
        return true;
    }

    /// <summary>Returns the most recent result for <paramref name="fanId"/>, or null.</summary>
    public FanTestResult? GetResult(string fanId)
    {
        lock (_lock)
            return _runs.TryGetValue(fanId, out var run) ? run.Result : null;
    }

    /// <summary>Cancels a running test. Returns false if no test is running for this fan.</summary>
    public bool Cancel(string fanId)
    {
        lock (_lock)
        {
            if (!_runs.TryGetValue(fanId, out var run) || run.Result.Status != "running")
                return false;
            run.Cts.Cancel();
            return true;
        }
    }

    /// <summary>
    /// Returns slim progress snapshots for all currently-running tests.
    /// Called by WebSocketHub on every WS tick. Empty list when idle.
    /// </summary>
    public IReadOnlyList<FanTestProgress> GetActiveProgress()
    {
        lock (_lock)
        {
            return _runs.Values
                .Where(r => r.Result.Status == "running")
                .Select(r =>
                {
                    var last = r.Result.Steps.LastOrDefault();
                    return new FanTestProgress
                    {
                        FanId            = r.Result.FanId,
                        Status           = r.Result.Status,
                        StepsDone        = r.Result.Steps.Count,
                        StepsTotal       = r.Result.Options.Steps + 1, // steps + the 0% step
                        CurrentPct       = r.CurrentSpeedPct,
                        CurrentRpm       = last?.Rpm,
                        Steps            = r.Result.Steps.ToList(),
                        MinOperationalPct = r.Result.MinOperationalPct,
                    };
                })
                .ToList();
        }
    }

    // -----------------------------------------------------------------------
    // Sweep loop
    // -----------------------------------------------------------------------

    private async Task RunSweepAsync(string fanId, TestRun run,
        string previousMode, FanCurve? previousCurve,
        CancellationToken ct)
    {
        _log.LogInformation("Fan benchmark started: {FanId} ({Steps} steps, {SettleMs}ms settle)",
            fanId, run.Result.Options.Steps, run.Result.Options.SettleMs);
        try
        {
            var opts = run.Result.Options;
            // Build speed list: 0%, step%, 2*step%, …, 100%
            double stepSize = 100.0 / opts.Steps;
            var speeds = Enumerable.Range(0, opts.Steps + 1)
                .Select(i => Math.Min(100.0, Math.Round(i * stepSize, 1)))
                .ToList();

            foreach (var speedPct in speeds)
            {
                ct.ThrowIfCancellationRequested();

                // Set speed directly on hardware (bypasses FanService state)
                _hw.SetFanSpeed(fanId, speedPct);
                lock (_lock) run.CurrentSpeedPct = speedPct;

                // Wait for RPM to settle
                await Task.Delay(opts.SettleMs, ct);

                // Read RPM from the latest sensor snapshot (SensorWorker keeps it fresh)
                var rpm = ReadRpm(fanId, _sensors.Latest.Readings);
                var spinning = rpm.HasValue && rpm.Value >= opts.MinRpmThreshold;

                var step = new FanTestStep
                {
                    SpeedPct = speedPct,
                    Rpm      = rpm,
                    Spinning = spinning,
                };

                lock (_lock)
                {
                    run.Result.Steps.Add(step);

                    // Stall detection: first spinning step
                    if (spinning && run.Result.MinOperationalPct == null)
                        run.Result.MinOperationalPct = speedPct;

                    // Max RPM at full speed
                    if (speedPct >= 100 && rpm.HasValue)
                        run.Result.MaxRpm = rpm;
                }
            }

            lock (_lock)
            {
                run.Result.Status      = "completed";
                run.Result.CompletedAt = DateTimeOffset.UtcNow;
            }

            _log.LogInformation(
                "Fan benchmark completed: {FanId} — stall={MinOp}% maxRpm={MaxRpm}",
                fanId, run.Result.MinOperationalPct, run.Result.MaxRpm);
        }
        catch (OperationCanceledException)
        {
            lock (_lock)
            {
                run.Result.Status      = "cancelled";
                run.Result.CompletedAt = DateTimeOffset.UtcNow;
            }
            _log.LogInformation("Fan benchmark cancelled: {FanId}", fanId);
        }
        catch (Exception ex)
        {
            lock (_lock)
            {
                run.Result.Status      = "failed";
                run.Result.Error       = ex.Message;
                run.Result.CompletedAt = DateTimeOffset.UtcNow;
            }
            _log.LogWarning(ex, "Fan benchmark failed: {FanId}", fanId);
        }
        finally
        {
            _fans.UnlockFromTest(fanId);
            RestorePreviousMode(fanId, previousMode, previousCurve);
        }
    }

    // -----------------------------------------------------------------------
    // Helpers
    // -----------------------------------------------------------------------

    private static double? ReadRpm(string fanId, IReadOnlyList<SensorReading> readings)
        => readings.FirstOrDefault(r =>
                r.SensorType == SensorTypeValues.FanRpm &&
                (r.Id == $"{fanId}_rpm" || r.Id.StartsWith(fanId)))
            ?.Value;

    private void RestorePreviousMode(string fanId, string previousMode, FanCurve? previousCurve)
    {
        try
        {
            switch (previousMode)
            {
                case "curve" when previousCurve != null:
                    _fans.SetCurve(previousCurve);
                    break;
                case "auto":
                    _fans.SetAuto(fanId);
                    break;
                // "manual" — leave at the last speed the test set; user can adjust
            }
        }
        catch (Exception ex)
        {
            _log.LogWarning(ex, "Failed to restore fan mode after benchmark: {FanId}", fanId);
        }
    }

    // -----------------------------------------------------------------------
    // Internal state
    // -----------------------------------------------------------------------

    private sealed class TestRun
    {
        public FanTestResult Result { get; }
        public CancellationTokenSource Cts { get; } = new();
        public Task Task { get; set; } = Task.CompletedTask;
        public double CurrentSpeedPct { get; set; }

        public TestRun(FanTestResult result) => Result = result;
    }
}

File: backend/app/services/fan_test_service.py
from __future__ import annotations

import asyncio
import logging
from dataclasses import dataclass
from datetime import datetime, timezone
from threading import Lock
from typing import Literal

from pydantic import BaseModel, Field

from app.models.sensors import SensorReading, SensorType

logger = logging.getLogger(__name__)


class FanTestOptions(BaseModel):
    steps: int = Field(default=10, ge=2, le=20)
    settle_ms: int = Field(default=2500, ge=500, le=10000)
    min_rpm_threshold: float = Field(default=50.0, ge=0.0, le=5000.0)


class FanTestStep(BaseModel):
    speed_pct: float
    rpm: float | None
    spinning: bool


class FanTestResult(BaseModel):
    fan_id: str
    status: Literal["running", "completed", "cancelled", "failed"] = "running"
    started_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    completed_at: datetime | None = None
    steps: list[FanTestStep] = Field(default_factory=list)
    min_operational_pct: float | None = None
    max_rpm: float | None = None
    options: FanTestOptions = Field(default_factory=FanTestOptions)
    error: str | None = None


class FanTestProgress(BaseModel):
    fan_id: str
    status: str
    steps_done: int
    steps_total: int
    current_pct: float
    current_rpm: float | None
    steps: list[FanTestStep]
    min_operational_pct: float | None


@dataclass
class _TestRun:
    result: FanTestResult
    task: asyncio.Task | None = None
    current_pct: float = 0.0


class FanTestService:
    """Runs per-fan benchmark sweeps and exposes progress/results for API + WS."""

    def __init__(self, backend, sensor_service, fan_service) -> None:
        self._backend = backend
        self._sensor_service = sensor_service
        self._fan_service = fan_service
        self._runs: dict[str, _TestRun] = {}
        self._lock = Lock()

    async def try_start(self, fan_id: str, options: FanTestOptions) -> tuple[bool, str]:
        fan_ids = await self._backend.get_fan_ids()
        if fan_id not in fan_ids:
            return False, f"Fan '{fan_id}' not found"

        with self._lock:
            existing = self._runs.get(fan_id)
            if existing and existing.result.status == "running":
                return False, f"A test is already running for fan '{fan_id}'"

            result = FanTestResult(fan_id=fan_id, options=options)
            run = _TestRun(result=result)
            self._runs[fan_id] = run
            self._fan_service.lock_for_test(fan_id)
            run.task = asyncio.create_task(self._run_sweep(fan_id, run))
        return True, ""

    def get_result(self, fan_id: str) -> FanTestResult | None:
        with self._lock:
            run = self._runs.get(fan_id)
            if run is None:
                return None
            return run.result.model_copy(deep=True)

    def cancel(self, fan_id: str) -> bool:
        with self._lock:
            run = self._runs.get(fan_id)
            if run is None or run.result.status != "running" or run.task is None:
                return False
            run.task.cancel()
            return True

    def get_active_progress(self) -> list[FanTestProgress]:
        progress: list[FanTestProgress] = []
        with self._lock:
            for run in self._runs.values():
                if run.result.status != "running":
                    continue
                current_rpm = run.result.steps[-1].rpm if run.result.steps else None
                progress.append(
                    FanTestProgress(
                        fan_id=run.result.fan_id,
                        status=run.result.status,
                        steps_done=len(run.result.steps),
                        steps_total=run.result.options.steps + 1,
                        current_pct=run.current_pct,
                        current_rpm=current_rpm,
                        steps=[s.model_copy(deep=True) for s in run.result.steps],
                        min_operational_pct=run.result.min_operational_pct,
                    )
                )
        return progress

    async def shutdown(self) -> None:
        tasks: list[asyncio.Task] = []
        with self._lock:
            for run in self._runs.values():
                if run.task and not run.task.done():
                    run.task.cancel()
                    tasks.append(run.task)
        if tasks:
            await asyncio.gather(*tasks, return_exceptions=True)

    async def _run_sweep(self, fan_id: str, run: _TestRun) -> None:
        opts = run.result.options
        step_size = 100.0 / opts.steps
        speeds = [min(100.0, round(i * step_size, 1)) for i in range(opts.steps + 1)]

        try:
            for speed in speeds:
                await self._backend.set_fan_speed(fan_id, speed)
                with self._lock:
                    run.current_pct = speed

                await asyncio.sleep(opts.settle_ms / 1000.0)
                rpm = self._read_rpm(fan_id, self._sensor_service.latest)
                spinning = rpm is not None and rpm >= opts.min_rpm_threshold

                step = FanTestStep(speed_pct=speed, rpm=rpm, spinning=spinning)
                with self._lock:
                    run.result.steps.append(step)
                    if spinning and run.result.min_operational_pct is None:
                        run.result.min_operational_pct = speed
                    if speed >= 100.0 and rpm is not None:
                        run.result.max_rpm = rpm

            with self._lock:
                run.result.status = "completed"
                run.result.completed_at = datetime.now(timezone.utc)

        except asyncio.CancelledError:
            with self._lock:
                run.result.status = "cancelled"
                run.result.completed_at = datetime.now(timezone.utc)
            raise
        except Exception as exc:
            logger.exception("Fan benchmark failed for %s", fan_id)
            with self._lock:
                run.result.status = "failed"
                run.result.error = str(exc)
                run.result.completed_at = datetime.now(timezone.utc)
        finally:
            self._fan_service.unlock_from_test(fan_id)

    @staticmethod
    def _read_rpm(fan_id: str, readings: list[SensorReading]) -> float | None:
        target = f"{fan_id}_rpm"
        for r in readings:
            if r.sensor_type == SensorType.FAN_RPM and (r.id == target or r.id.startswith(fan_id)):
                return r.value
        return None
